src/Raven.Server/Storage/StorageSpaceMonitor.cs

```diff
@@ -0,0 +1,221 @@
+            _logger = LoggingSource.Instance.GetLogger(database.Name, GetType().FullName);
+
+            _timer = new Timer(Run, null, CheckFrequency, CheckFrequency);
+        }
+
+        internal void Run(object state)
+        {
+            if (_notificationCenter.IsInitialized == false)
+                return;
+
+            if (Monitor.TryEnter(_lock) == false)
+                return;
+
+            try
+            {
+                List<LowDiskSpace> lowSpaceDisks = null;
+
+                var repeat = false;
+
+                do
+                {
+                    var didCleanup = false;
+
+                    foreach (var lowSpaceDisk in GetLowSpaceDisks())
+                    {
+                        if (repeat == false)
+                        {
+                            foreach (var storageEnvironment in lowSpaceDisk.Value)
+                            {
+                                storageEnvironment.Cleanup(deleteRecyclableJournals: true);
+                            }
+
+                            didCleanup = true;
+                        }
+                        else
+                        {
+                            if (lowSpaceDisks == null)
+                                lowSpaceDisks = new List<LowDiskSpace>();
+
+                            lowSpaceDisks.Add(lowSpaceDisk.Key);
+
+                            repeat = false;
+                        }
+
+                        if (didCleanup)
+                            repeat = true;
+                    }
+
+                } while (repeat);
+
+                if (lowSpaceDisks != null && lowSpaceDisks.Count > 0)
+                {
+                    var message =
+                        $"The following disks are running out of space:{Environment.NewLine} {string.Join(Environment.NewLine, lowSpaceDisks)}";
+
+                    _notificationCenter.Add(AlertRaised.Create(_database.Name, "Low free disk space", message, AlertType.LowDiskSpace, NotificationSeverity.Warning,
+                        "low-disk-space"));
+
+                    if (_logger.IsOperationsEnabled)
+                        _logger.Operations(message);
+                }
+            }
+            catch (Exception e)
+            {
+                if (_logger.IsOperationsEnabled)
+                    _logger.Operations("Failed to run storage space monitor", e);
+            }
+            finally
+            {
+                Monitor.Exit(_lock);
+            }
+        }
+
+        private Dictionary<LowDiskSpace, List<StorageEnvironment>> GetLowSpaceDisks()
+        {
+            var result = new Dictionary<LowDiskSpace, List<StorageEnvironment>>();
+
+            var drives = DriveInfo.GetDrives();
+            
+            foreach (var item in _database.GetAllStoragesEnvironment())
+            {
+                var dataDrive = DiskSpaceChecker.GetFreeDiskSpace(item.Environment.Options.BasePath.FullPath, drives);
```

Comment:
Remember that user may also use  junction points, need to verify this for the `Journals` and `Temp` directory as well inside here.

Reply:
`Temp` is checked below. But `Journals` needs to be checked separately

Reply:
`Temp` is checked only if it is specially declared, but you can also setup junction point there too

Reply:
Just to be sure, does it mean that we need to call `GetDiskFreeSpaceEx`:
https://github.com/ravendb/ravendb/blob/v4.1/src/Raven.Server/Utils/DiskSpaceChecker.cs#L14
for `Journals` and `Temp` directories for each environment we have just in case it's a link?

Reply:
GetDiskFreeSpaceEx doesn't provide the free disk space if it's a link.

http://issues.hibernatingrhinos.com/issue/RavenDB-12010

Reply:
The docs says it does: https://docs.microsoft.com/en-us/windows/desktop/fileio/symbolic-link-effects-on-file-systems-functions#getdiskfreespaceex

Reply:
@arekpalinski And yes, I think we need to check if each of them is a junction.
@grisha-kotler Note that we also need this on Linux

Reply:
`GetDiskFreeSpaceEx` seems to be working as expected - http://issues.hibernatingrhinos.com/issue/RavenDB-12010#focus=streamItem-67-31686-0-0

Reply:
Always checking temp and journal dirs as well added in: https://github.com/ravendb/ravendb/pull/7850/commits/a6dcded55dee85d7ca721022304352348ffe3a32

Handling of symbolic links and proper drive detection then will be done by @grisha-kotler (RavenDB-12010)